File: Doubling Season/TokenDefinition.swift
//
//  TokenDefinition.swift
//  Doubling Season
//
//  Created on 10/10/25.
//

import Foundation
import SwiftData

/// Represents a token definition from the TokenDatabase.json file
/// This struct is used to decode and work with token data in the app
struct TokenDefinition: Codable, Identifiable, Hashable {
    // MARK: - Properties
    
    /// The name of the token
    let name: String
    
    /// The abilities text of the token
    let abilities: String
    
    /// The power/toughness of the token (e.g., "2/2", "*/*")
    let pt: String
    
    /// The color identity of the token (e.g., "W", "UB", "WUBRG")
    let colors: String
    
    /// The type line of the token (e.g., "Token Creature — Zombie")
    let type: String
    
    // MARK: - Identifiable

    /// Unique identifier for SwiftUI list operations
    var id: String {
        // Create a composite ID to handle tokens with the same name but different stats/abilities
        return "\(name)|\(pt)|\(colors)|\(type)|\(abilities)"
    }
    
    // MARK: - Computed Properties
    
    /// Searchable text that combines all relevant fields for search functionality
    var searchableText: String {
        // Combine all text fields for comprehensive search
        return "\(name) \(abilities) \(type) \(colors) \(pt)".lowercased()
    }
    
    /// Returns true if this token is a creature (has power/toughness)
    var isCreature: Bool {
        return !pt.isEmpty
    }
    
    /// Returns a formatted color string for display
    var colorDisplay: String {
        if colors.isEmpty {
            return "Colorless"
        } else if colors.count == 1 {
            switch colors {
            case "W": return "White"
            case "U": return "Blue"
            case "B": return "Black"
            case "R": return "Red"
            case "G": return "Green"
            default: return colors
            }
        } else if colors == "WUBRG" || colors == "BGRUW" {
            return "All Colors"
        } else {
            // Multi-color
            return "Multicolor (\(colors))"
        }
    }
    
    /// Returns a clean type string without "Token" prefix
    var cleanType: String {
        if type.hasPrefix("Token ") {
            return String(type.dropFirst(6))
        }
        return type
    }
    
    // MARK: - Methods
    
    /// Converts this TokenDefinition to an Item for creation in the game
    /// - Parameters:
    ///   - amount: The number of tokens to create
    ///   - createTapped: Whether the tokens should enter tapped
    /// - Returns: An Item instance ready to be added to the game
    func toItem(amount: Int = 1, createTapped: Bool = false) -> Item {
        return Item(
            abilities: abilities,
            name: name,
            pt: pt,
            colors: colors,
            amount: amount,
            createTapped: createTapped
        )
    }
    
    /// Checks if this token matches a search query
    /// - Parameter query: The search string to match against
    /// - Returns: true if the token matches the search query
    func matches(searchQuery query: String) -> Bool {
        if query.isEmpty {
            return true
        }
        
        let lowercasedQuery = query.lowercased()
        
        // Check for exact matches in important fields
        if name.lowercased().contains(lowercasedQuery) {
            return true
        }
        
        // Check searchable text for partial matches
        return searchableText.contains(lowercasedQuery)
    }
    
    // MARK: - Hashable
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(name)
    }
    
    // MARK: - Equatable
    
    static func == (lhs: TokenDefinition, rhs: TokenDefinition) -> Bool {
        return lhs.name == rhs.name
    }
}

// MARK: - Extensions

extension TokenDefinition {
    /// Common token categories for filtering
    enum Category: String, CaseIterable {
        case creature = "Creature"
        case artifact = "Artifact"
        case enchantment = "Enchantment"
        case emblem = "Emblem"
        case dungeon = "Dungeon"
        case counter = "Counter"
        case other = "Other"
        
        var displayName: String {
            return self.rawValue
        }
    }
    
    /// Returns the primary category of this token
    var category: Category {
        let typeLC = type.lowercased()
        
        if typeLC.contains("emblem") {
            return .emblem
        } else if typeLC.contains("dungeon") {
            return .dungeon
        } else if typeLC.contains("counter") {
            return .counter
        } else if typeLC.contains("creature") {
            return .creature
        } else if typeLC.contains("artifact") {
            return .artifact
        } else if typeLC.contains("enchantment") {
            return .enchantment
        } else {
            return .other
        }
    }
}

File: lib/models/token_definition.dart
import 'item.dart';

class ArtworkVariant {
  final String set;
  final String url;

  ArtworkVariant({required this.set, required this.url});

  factory ArtworkVariant.fromJson(Map<String, dynamic> json) {
    return ArtworkVariant(
      set: json['set'] as String? ?? '',
      url: json['url'] as String? ?? '',
    );
  }
}

class TokenDefinition {
  final String name;
  final String abilities;
  final String pt;
  final String colors;
  final String type;
  final int popularity;
  final List<ArtworkVariant> artwork;

  TokenDefinition({
    required this.name,
    required this.abilities,
    required this.pt,
    required this.colors,
    required this.type,
    required this.popularity,
    this.artwork = const [],
  });

  // CRITICAL: Composite ID must match deduplication logic in process_tokens.py
  String get id => '$name|$pt|$colors|$type|$abilities';

  factory TokenDefinition.fromJson(Map<String, dynamic> json) {
    // Parse artwork array
    final artworkJson = json['artwork'] as List<dynamic>?;
    final artworkList = artworkJson
            ?.map((art) => ArtworkVariant.fromJson(art as Map<String, dynamic>))
            .toList() ??
        [];

    return TokenDefinition(
      name: json['name'] as String? ?? '',
      abilities: json['abilities'] as String? ?? '',
      pt: json['pt'] as String? ?? '',
      colors: json['colors'] as String? ?? '',
      type: json['type'] as String? ?? '',
      popularity: json['popularity'] as int? ?? 0,
      artwork: artworkList,
    );
  }

  bool matches({required String searchQuery}) {
    if (searchQuery.isEmpty) return true;
    final query = searchQuery.toLowerCase();
    return name.toLowerCase().contains(query) ||
        abilities.toLowerCase().contains(query) ||
        pt.toLowerCase().contains(query) ||
        type.toLowerCase().contains(query);
  }

  Item toItem({required int amount, required bool createTapped}) {
    return Item(
      name: name,
      pt: pt,
      abilities: abilities,
      colors: colors,
      type: type,
      amount: amount,
      tapped: createTapped ? amount : 0,
      summoningSick: amount, // Always apply summoning sickness to new tokens
    );
  }

  String get cleanType {
    // Remove "Token" suffix if present
    return type.replaceAll(RegExp(r'\s+Token$', caseSensitive: false), '');
  }

  // Category for filtering
  Category get category {
    final lowerType = type.toLowerCase();
    if (lowerType.contains('creature')) return Category.creature;
    if (lowerType.contains('artifact')) return Category.artifact;
    if (lowerType.contains('enchantment')) return Category.enchantment;
    if (lowerType.contains('emblem')) return Category.emblem;
    if (lowerType.contains('dungeon')) return Category.dungeon;
    if (name.toLowerCase().contains('counter')) return Category.counter;
    return Category.other;
  }
}

enum Category {
  creature,
  artifact,
  enchantment,
  emblem,
  dungeon,
  counter,
  other;

  String get displayName {
    switch (this) {
      case Category.creature:
        return 'Creature';
      case Category.artifact:
        return 'Artifact';
      case Category.enchantment:
        return 'Enchantment';
      case Category.emblem:
        return 'Emblem';
      case Category.dungeon:
        return 'Dungeon';
      case Category.counter:
        return 'Counter';
      case Category.other:
        return 'Other';
    }
  }
}
